examples/step-37/step-37.cc

```diff
@@ -168,228 +170,137 @@ namespace Step37
+  // coefficiient values.
+  //
+  // This program makes use of the data cache for finite element operator
+  // application that is integrated in deal.II. This data cache class is
+  // called MatrixFree. It contains mapping information (Jacobians) and index
+  // relations between local and global degrees of freedom. It also contains
+  // constraints like the ones from hanging nodes or Dirichlet boundary
+  // conditions. Moreover, it can issue a loop over all cells in %parallel,
+  // making sure that only cells are worked on that do not share any degree of
+  // freedom (this makes the loop thread-safe when writing into destination
+  // vectors). This is a more advanced strategy compared to the WorkStream
+  // class described in the @ref threads module. Of course, to not destroy
+  // thread-safety, we have to be careful when writing into class-global
+  // structures.
   //
-  // In this class, we store the actual MatrixFree object, the variable
-  // coefficient that is evaluated at all quadrature points (so that we don't
-  // have to recompute it during matrix-vector products), and a vector that
-  // contains the diagonal of the matrix that we need for the multigrid
-  // smoother. We choose to let the user provide the diagonal in this program,
-  // but we could also integrate a function in this class to evaluate the
-  // diagonal. Unfortunately, this forces us to define matrix entries at two
-  // places, once when we evaluate the product and once for the diagonal, but
-  // the work is still much less than when we compute sparse matrices.
+  // The class implementing the Laplace operator has three template arguments,
+  // one for the dimension (as many deal.II classes carry), one for the degree
+  // of the finite element (which we need to enable efficient computations
+  // through the FEEvaluation class), and one for the underlying scalar
+  // type. We want to use <code>double</code> numbers (i.e., double precision,
+  // 64-bit floating point) for the final matrix, but floats (single
+  // precision, 32-bit floating point numbers) for the multigrid level
+  // matrices (as that is only a preconditioner, and floats can be processed
+  // twice as fast).
   //
   // As a sidenote, if we implemented several different operations on the same
   // grid and degrees of freedom (like a mass matrix and a Laplace matrix), we
-  // would have to have two classes like the current one for each of the
-  // operators (maybe with a common base class). However, in that case, we
-  // would not store a MatrixFree object in this class to avoid doing the
-  // expensive work of precomputing everything MatrixFree stores
-  // twice. Rather, we would keep this object in the main class and simply
-  // store a reference.
+  // would define two classes like the current one for each of the operators
+  // (maybe with a common base class), and let both of them refer to the same
```

Comment:
`maybe with a common base class` -> `derived from the MatrixFreeOperators::Base`.